File: 04 April/04-30/binaryCrossword.js
function isInCrossword(char) {
  const GRID = [
    [0, 1, 0, 0, 0, 0, 0, 1],
    [0, 1, 1, 0, 1, 1, 1, 1],
    [0, 1, 0, 0, 0, 1, 0, 0],
    [0, 1, 1, 0, 0, 1, 0, 1],
    [0, 1, 0, 1, 0, 0, 1, 0],
    [0, 1, 0, 1, 0, 1, 0, 0],
    [0, 1, 1, 0, 1, 0, 0, 0],
    [1, 0, 1, 0, 1, 1, 1, 0],
  ];
  let gridAllPossibleNumbers = [];

  for (const row of GRID) {
    gridAllPossibleNumbers = gridAllPossibleNumbers.concat([row.join(""), row.toReversed().join("")]);
  }

  for (let j = 0; j < GRID[0].length; ++j) {
    let number = "";
    let reversedNumber = "";

    for (const row of GRID) {
      number += row[j];
      reversedNumber = row[j] + reversedNumber;
    }

    gridAllPossibleNumbers = gridAllPossibleNumbers.concat([number, reversedNumber]);
  }

  return gridAllPossibleNumbers.includes(char.charCodeAt(0).toString(2).padStart(8, 0));
}

// --- TEST SUITE ---

const testsText = String.raw`
1. isInCrossword("I") should return true.
2. isInCrossword("D") should return true.
3. isInCrossword("0") should return true.
4. isInCrossword("u") should return true.
5. isInCrossword("Y") should return false.
6. isInCrossword("p") should return false.
7. isInCrossword("1") should return false.
8. isInCrossword("Q") should return false.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(`(${test.output})`);

    if (JSON.stringify(functionCallOutput) === JSON.stringify(testOutput)) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${JSON.stringify(testOutput)}\nGot: ${JSON.stringify(functionCallOutput)}`);

      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

runTests(testData);

File: 04 April/04-30/binary_crossword.py
def is_in_crossword(char):
    GRID = [
        [0, 1, 0, 0, 0, 0, 0, 1],
        [0, 1, 1, 0, 1, 1, 1, 1],
        [0, 1, 0, 0, 0, 1, 0, 0],
        [0, 1, 1, 0, 0, 1, 0, 1],
        [0, 1, 0, 1, 0, 0, 1, 0],
        [0, 1, 0, 1, 0, 1, 0, 0],
        [0, 1, 1, 0, 1, 0, 0, 0],
        [1, 0, 1, 0, 1, 1, 1, 0],
    ]
    grid_all_possible_numbers = []

    for row in GRID:
        grid_all_possible_numbers.extend(["".join(map(str, row)), "".join(map(str, reversed(row)))])

    for j in range(len(GRID[0])):
        number = "".join([str(row[j]) for row in GRID])

        grid_all_possible_numbers.extend([number, number[::-1]])

    return any(number == "{:08b}".format(ord(char)) for number in grid_all_possible_numbers)

# --- TEST SUITE ---

tests_text = r'''
1. is_in_crossword("I") should return True.
2. is_in_crossword("D") should return True.
3. is_in_crossword("0") should return True.
4. is_in_crossword("u") should return True.
5. is_in_crossword("Y") should return False.
6. is_in_crossword("p") should return False.
7. is_in_crossword("1") should return False.
8. is_in_crossword("Q") should return False.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)